satpy/readers/viirs_sdr.py

```diff
@@ -219,18 +219,29 @@ def get_file_units(self, dataset_id, ds_info):
             LOG.debug("Unknown units for file key '%s'", dataset_id)
         return file_units
 
-    def scale_swath_data(self, data, scaling_factors):
+    def scale_swath_data(self, data, scaling_factors, dataset_group):
         """Scale swath data using scaling factors and offsets.
 
         Multi-granule (a.k.a. aggregated) files will have more than the usual two values.
         """
-        num_grans = len(scaling_factors) // 2
-        gran_size = data.shape[0] // num_grans
+        rows_per_gran = self._get_rows_per_granule(dataset_group)
         factors = scaling_factors.where(scaling_factors > -999, np.float32(np.nan))
-        factors = factors.data.reshape((-1, 2))
-        factors = xr.DataArray(da.repeat(factors, gran_size, axis=0),
-                               dims=(data.dims[0], 'factors'))
-        data = data * factors[:, 0] + factors[:, 1]
+        factors = factors.data.reshape((-1, 2)).rechunk((1, 2))  # make it so map_blocks happens per factor
+        data = self._map_and_apply_factors(data, factors, rows_per_gran)
+        return data
+
+    @staticmethod
+    def _map_and_apply_factors(data, factors, rows_per_gran):
+        # The user may have requested a different chunking scheme, but we need
+        # per granule chunking right now so factor chunks map 1:1 to data chunks
```

Comment:
Just a general comment here: I think the chunking scheme should be dependent of the data, not directly on what the user requested. So I'd like to see satpy accepting chunksizes in MB for example instead of number of pixels, as it could be wasteful. The developers of the filehandler know probably best how to chunk the data for size constraints.

Reply:
I agree, but is that something I should try to fix in this PR? This reader uses the HDF5 utility handler so it would be difficult to do it cleanly for just this reader.

Reply:
Nothing you need to fix here. I was just reacting to the comment.